mmocr/models/textrecog/decoders/master_decoder.py

```diff
@@ -0,0 +1,232 @@
+
+    d_k = value.size(-1)
+    score = torch.matmul(query, key.transpose(-2, -1) / math.sqrt(d_k))
+    if mask is not None:
+        #  score = score.masked_fill(mask == 0, -1e9)  # b, h, L, L
+        score = score.masked_fill(mask == 0, -6.55e4)  # for fp16
+    p_attn = F.softmax(score, dim=-1)
+    if dropout is not None:
+        p_attn = dropout(p_attn)
+    return torch.matmul(p_attn, value), p_attn
+
+
+class MultiHeadAttention(nn.Module):
+
+    def __init__(self, headers, d_model, dropout):
+        super(MultiHeadAttention, self).__init__()
+
+        assert d_model % headers == 0
+        self.d_k = int(d_model / headers)
+        self.headers = headers
+        self.linears = clones(nn.Linear(d_model, d_model), 4)
+        self.attn = None
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, query, key, value, mask=None):
+        nbatches = query.size(0)
+        # 1) Do all the linear projections in batch from d_model => h x d_k
+        query, key, value = \
+            [linear(x).view(nbatches,
+                            -1,
+                            self.headers,
+                            self.d_k
+                            ).transpose(1, 2)
+             for linear, x in zip(self.linears, (query, key, value))
+             ]
+        # 2) Apply attention on all the projected vectors in batch
+        x, self.attn = self_attention(
+            query, key, value, mask=mask, dropout=self.dropout)
+        x = x.transpose(1, 2).contiguous().view(nbatches, -1,
+                                                self.headers * self.d_k)
+        return self.linears[-1](x)
+
+
+class DecoderLayer(nn.Module):
+    """Decoder is made of self attention, source attention and feed forward."""
+
+    def __init__(self, size, self_attn, src_attn, feed_forward, dropout):
+        super(DecoderLayer, self).__init__()
+        self.size = size
+        self.self_attn = MultiHeadAttention(**self_attn)
+        self.src_attn = MultiHeadAttention(**src_attn)
+        self.feed_forward = FeedForward(**feed_forward)
+        self.sublayer = clones(SubLayerConnection(size, dropout), 3)
+
+    def forward(self, x, feature, src_mask, tgt_mask):
+        x = self.sublayer[0](x, lambda x: self.self_attn(x, x, x, tgt_mask))
+        x = self.sublayer[1](
+            x, lambda x: self.src_attn(x, feature, feature, src_mask))
+        return self.sublayer[2](x, self.feed_forward)
+
+
+@DECODERS.register_module()
+class MasterDecoder(BaseDecoder):
+    """Decoder module in `MASTER.
+
+    <https://arxiv.org/abs/1910.02562>`_.
+
+    Args:
+        N (int): Number of transformer layer in decoder.
+        decoder (dict): Transformer decoder setting config dict.
+        d_model (int): Output channel of decoder.
+        num_classes (int): Output class number.
+        start_idx (int): Index of start token.
+        padding_idx (int): Index of padding token.
+        max_seq_len (int): Maximum sequence length for decoding.
+    """
```

Comment:
Hi，when I use BaseTransformerLayer in MMOCR， I have meet this error.

Traceback (most recent call last):
  File "tools/train.py", line 230, in <module>
    main()
  File "tools/train.py", line 226, in main
    meta=meta)
  File "/data_0/mmocr/mmocr/apis/train.py", line 155, in train_detector
    runner.run(data_loaders, cfg.workflow)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/runner/epoch_based_runner.py", line 127, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/runner/epoch_based_runner.py", line 50, in train
    self.run_iter(data_batch, train_mode=True, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/runner/epoch_based_runner.py", line 30, in run_iter
    **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/parallel/data_parallel.py", line 67, in train_step
    return self.module.train_step(*inputs[0], **kwargs[0])
  File "/data_0/mmocr/mmocr/models/textrecog/recognizer/base.py", line 158, in train_step
    losses = self(**data)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/runner/fp16_utils.py", line 97, in new_func
    return old_func(*args, **kwargs)
  File "/data_0/mmocr/mmocr/models/textrecog/recognizer/base.py", line 84, in forward
    return self.forward_train(img, img_metas, **kwargs)
  File "/data_0/mmocr/mmocr/models/textrecog/recognizer/encode_decode_recognizer.py", line 110, in forward_train
    feat, out_enc, targets_dict, img_metas, train_mode=True)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/data_0/mmocr/mmocr/models/textrecog/decoders/master_decoder.py", line 200, in forward
    return self.forward_train(feat, out_enc, targets_dict, img_metas)
  File "/data_0/mmocr/mmocr/models/textrecog/decoders/master_decoder.py", line 174, in forward_train
    return self.decode(padded_targets[:, :-1], out_enc, src_mask, tgt_mask)
  File "/data_0/mmocr/mmocr/models/textrecog/decoders/master_decoder.py", line 141, in decode
    attn_masks=attn_masks)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/cnn/bricks/transformer.py", line 489, in forward
    **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/utils/misc.py", line 330, in new_func
    output = old_func(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/cnn/bricks/transformer.py", line 206, in forward
    key_padding_mask=key_padding_mask)[0]
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/mmcv/cnn/bricks/transformer.py", line 119, in forward_wrapper
    attn_output, attn_output_weights = forward(**kwargs)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/activation.py", line 987, in forward
    attn_mask=attn_mask)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py", line 4636, in multi_head_attention_forward
    q, k, v = linear(query, in_proj_weight, in_proj_bias).chunk(3, dim=-1)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py", line 1753, in linear
    return torch._C._nn.linear(input, weight, bias)
RuntimeError: Tensor for argument #3 'mat2' is on CPU, but expected it to be on GPU (while checking arguments for addmm)

Is it occur because nn.linear in mmcv's BaseTransformerLayer? I am sure my input tensor of BaseTransformerLayer is on the same device.

Reply:
I set batch_first=True in BaseTransformerLayer. And I have found this [issue](https://github.com/open-mmlab/mmcv/pull/1418) in mmcv(version 1.3.8), but it seems not work on my situation.

Reply:
@JiaquanYe  Sorry for the late reply. What's your MMCV version? It should have been fixed after 1.3.17.

Reply:
> @JiaquanYe Sorry for the late reply. What's your MMCV version? It should have been fixed after 1.3.17.

My mmcv version is v1.3.8. I will try again after update mmcv to 1.3.17. I haved permuted the tensor dimension order before forward BaseTransformerLayer, the error disappear. 

In original MASTER code, attn_mask will be cut by gt label via the code in make_mask func:
trg_pad_mask = (tgt != self.PAD).unsqueeze(1).unsqueeze(3).byte()

It means the attn_masks in a mini-batch is not the same.

When I use BaseTransformerLayer, I pass a 3D attn_mask, tensor like (bs, max_length, max_length), it will raise an Error:
RuntimeError: The size of the 3D attn_mask is not correct.

It seems 3D attn_mask is not support. Is  there any way to deal with this situation?

Reply:
Could you post the error trace log here? It seems MMCV's wrapper eventually invokes `nn.MultiheadAttention` which supports 3D masks. I'm not sure where it went wrong.

Reply:
Yes, I am sure nn.MultiheadAttention supports 3D mask, after I check the MMCV's source code.
BTW, everything go well in BaseTransformerLayer after I update MMCV 1.3.8 to MMCV 1.3.18, 'batch_first' is supported in this version.